code/modules/cm_aliens/XenoStructures.dm

```diff
@@ -844,6 +845,275 @@
+		if(cur_hive_num == hive_number)
+			xeno_announcement(SPAN_XENOANNOUNCE("The King is growing at [get_area_name(loc)]. Protect it at all costs!"), cur_hive_num, XENO_GENERAL_ANNOUNCE)
+		else
+			xeno_announcement(SPAN_XENOANNOUNCE("Another hive's King is growing at [get_area_name(loc)]."), cur_hive_num, XENO_GENERAL_ANNOUNCE)
+
+/obj/effect/alien/resin/king_cocoon/proc/check_pylons()
+	var/datum/hive_status/hive = GLOB.hive_datum[XENO_HIVE_NORMAL]
+
+	if(length(hive.active_endgame_pylons) < 2)
+		qdel(src)
+		return
+
+	if(chosen_candidate || rolling_candidates)
+		return
+
+	if(icon_state == "hatching" || icon_state == "hatched")
+		return
+
+	var/groundside_humans = 0
+	for(var/mob/living/carbon/human/current_human as anything in GLOB.alive_human_list)
+		if(!(isspecieshuman(current_human) || isspeciessynth(current_human)))
+			continue
+
+		var/turf/turf = get_turf(current_human)
+		if(is_ground_level(turf?.z))
+			groundside_humans += 1
+			
+			if(groundside_humans > 12)
+				return
+
+	deltimer(timer)
+
+	choose_candidate(TRUE)
+
+/obj/effect/alien/resin/king_cocoon/proc/start_growing()
+	icon_state = "growing"
+	timer = addtimer(CALLBACK(src, PROC_REF(announce_halfway)), 5 MINUTES, TIMER_UNIQUE|TIMER_STOPPABLE|TIMER_DELETE_ME)
+
+/obj/effect/alien/resin/king_cocoon/proc/announce_halfway()
+	timer = addtimer(CALLBACK(src, PROC_REF(choose_candidate)), 4 MINUTES, TIMER_UNIQUE|TIMER_STOPPABLE|TIMER_DELETE_ME)
+
+	marine_announcement("ALERT.\n\nUNSUAL ENERGY BUILDUP DETECTED IN [get_area_name(loc)].\n\nESTIMATED TIME UNTIL COMPLETION - 5 MINUTES.", "[MAIN_AI_SYSTEM] Biological Scanner", 'sound/misc/notice1.ogg')
+	var/datum/hive_status/hive
+	for(var/cur_hive_num in GLOB.hive_datum)
+		hive = GLOB.hive_datum[cur_hive_num]
+		if(!length(hive.totalXenos))
+			continue
+		if(cur_hive_num == hive_number)
+			xeno_announcement(SPAN_XENOANNOUNCE("The King will hatch in approximately 5 minutes."), cur_hive_num, XENO_GENERAL_ANNOUNCE)
+		else
+			xeno_announcement(SPAN_XENOANNOUNCE("Another hive's King will hatch in approximately 5 minutes."), cur_hive_num, XENO_GENERAL_ANNOUNCE)
+
+#define KING_PLAYTIME_HOURS (50 HOURS)
+
+/obj/effect/alien/resin/king_cocoon/proc/is_candidate_valid(datum/hive_status/hive, mob/candidate, playtime_restricted = TRUE, skip_playtime=TRUE)
+	if(!candidate.client)
+		return FALSE
+	if(isfacehugger(candidate) || islesserdrone(candidate))
+		return FALSE
+	if(playtime_restricted)
+		if(candidate.client.get_total_xeno_playtime() < KING_PLAYTIME_HOURS)
+			return FALSE
+	else if(candidate.client.get_total_xeno_playtime() >= KING_PLAYTIME_HOURS && skip_playtime)
+		return FALSE // We do this under the assumption we tried it the other way already so don't ask twice
+	for(var/mob_name in hive.banished_ckeys)
+		if(hive.banished_ckeys[mob_name] == candidate.ckey)
+			return FALSE
+	return TRUE
+
+/obj/effect/alien/resin/king_cocoon/proc/try_roll_candidate(datum/hive_status/hive, mob/candidate, playtime_restricted = TRUE)
```

Comment:
```suggestion
/**
 * Returns TRUE if a valid candidate accepts a TGUI alert asking them to be King.
 * 
 * Arguments:
 * * hive: The hive_status to check banished ckeys against
 * * candidate: The mob that we want to ask
 * * playtime_restricted: Determines whether being below KING_PLAYTIME_HOURS makes the candidate invalid (otherwise above)
 */
/obj/effect/alien/resin/king_cocoon/proc/try_roll_candidate(datum/hive_status/hive, mob/candidate, playtime_restricted = TRUE)
```